test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Reader/JsonLight/PropertyAndValueJsonLightReaderIntegrationTests.cs

```diff
@@ -773,6 +845,246 @@ public void ReadingNullValueForDeclaredCollectionPropertyInComplexTypeShouldFail
         }
 
+        [Fact]
+        public void ReadingPropertyInTopLevelOrInComplexTypeShouldRestoreOmittedNullValues()
+        {
+            EdmEntityType entityType;
+            EdmEntitySet entitySet;
+            EdmModel model = BuildEdmModelForOmittedNullValuesTestCases(out entityType, out entitySet);
+
+            const string payload =@"{
+                    ""@odata.context"":""http://www.example.com/$metadata#EntityNs.MyContainer.People/$entity"",
+                    ""@odata.id"":""http://mytest"",
+                    ""Id"":0,
+                    ""Education"":{""Id"":1}
+                }";
+
+            List<ODataResource> entries = new List<ODataResource>();
+            this.ReadEntryPayload(model, payload, entitySet, entityType,
+                reader =>
+                {
+                    switch (reader.State)
+                    {
+                        case ODataReaderState.ResourceStart:
+                            entries.Add(reader.Item as ODataResource);
+                            break;
+                        default:
+                            break;
+                    }
+                },
+                nullValuesOmitted: true);
+
+            Assert.Equal(2, entries.Count);
+
+            ODataResource edu = entries.FirstOrDefault(s => string.Equals(s.TypeName, "NS.Edu", StringComparison.Ordinal));
+            edu.Should().NotBeNull();
+            edu.Properties.FirstOrDefault(s => string.Equals(s.Name, "Id", StringComparison.Ordinal)).Value.Should().Be(1);
+            // null-able collection should be restored.
+            edu.Properties.FirstOrDefault(s => string.Equals(s.Name, "Campuses", StringComparison.Ordinal)).Value.Should().BeNull();
+            // null-able property value should be restored.
+            edu.Properties.FirstOrDefault(s => string.Equals(s.Name, "SchoolName", StringComparison.Ordinal)).Value.Should().BeNull();
+
+            ODataResource person = entries.FirstOrDefault(s => string.Equals(s.TypeName, "NS.Person", StringComparison.Ordinal));
+            person.Should().NotBeNull();
+            person.Properties.FirstOrDefault(s => string.Equals(s.Name, "Id", StringComparison.Ordinal)).Value.Should().Be(0);
+            // omitted null-able property should be restored as null.
+            person.Properties.FirstOrDefault(s => string.Equals(s.Name, "Height", StringComparison.Ordinal)).Value.Should().BeNull();
+            // missing non-null-able property should not be restored.
+            person.Properties.Any(s => string.Equals(s.Name, "Weight", StringComparison.Ordinal)).Should().BeFalse();
+        }
+
+        [Fact]
+        public void ReadingPropertyInTopLevelOrInComplexTypeShouldRestoreOmittedNullValuesWithSelectedPropertiesEntireSubTree()

```

Comment:
Add a test case for dynamic property selected, and it is not in the payload, should be materialized as null.

Reply:
Add unknown properties to context URI in test cases to verify unknown properties are ignored during deserialization.
Dynamic properties don't show up in context URI in response, therefore omitted one are not materialized as nulls.

---
In reply to: [173008510](https://github.com/OData/odata.net/pull/1106#discussion_r173008510) [](ancestors = 173008510)

Reply:
This comment seems wrong. Dynamic properties do show up in ContextUrl if explicitly selected.

---
In reply to: [174266340](https://github.com/OData/odata.net/pull/1106#discussion_r174266340) [](ancestors = 174266340,173008510)